### ISSUE TITLE
Storing comments (form data) in long life cookies

### ISSUE THREAD

**Issue by caffeineinc** [silverstripe/silverstripe-comments#263]
# Background
The module currently stores the form data from a comment submission into a cookie. I think the intent was that if the comment failed validation the form could be pre-populated with that content.  This makes more sense with the optional form elements for Name / URL / Email information, across the site but the Comment data is also stored. 

Storing this data across the site, is also dubious as it would remain for people on public computers for as long as the default cookie timeout (default to 90 days without configuration). 

class CommentForm:L135
```
        // load any data from the cookies
        if ($data = Cookie::get('CommentsForm_UserData')) {
            $data = Convert::json2array($data);

            $this->loadDataFrom(array(
                'Name'  => isset($data['Name']) ? $data['Name'] : '',
                'URL'   => isset($data['URL']) ? $data['URL'] : '',
                'Email' => isset($data['Email']) ? $data['Email'] : ''
            ));

            // allow previous value to fill if comment not stored in cookie (i.e. validation error)
            $prevComment = Cookie::get('CommentsForm_Comment');

            if ($prevComment && $prevComment != '') {
                $this->loadDataFrom(array('Comment' => $prevComment));
            }
        }
```

When posting  the user data is 'cached':

  ```
// cache users data
        Cookie::set('CommentsForm_UserData', Convert::raw2json($data));
        Cookie::set('CommentsForm_Comment', $data['Comment']);
```

but not all of these cookies aren't cleaned up and whatever the form data was (CommentsForm_UserData), the cookie remains for every request, which 
- prevents caching of the page under certain scenarios.
- means the form content is persisted in other ways, even if the comment wasn't accepted by the system. 
- it also is limited by the amount of information you can store in a cookie. 

## Issues: 
If a user were to add XSS to a form comment, then that XSS is present in the cookie data, which i don't think is exploitable, but can be mistaken as a serious issue.

Saving data to the cookie is persisted for a long time (90 days), which might be a privacy issue.

I would make a suggestion to use form session instead and clear the submission contents if the write was successful.

This also has other issues, as the keys aren't comment specific, so can be applied to any comment form on the site.

Using cookies in this way we can't really overload the behaviour or prevent it without forking it. 

## Resolution
I think a better way might be to set the formdata to the session, as one can't overload the cookie data after they're set (we can force expiry but the data still goes to the browser.

**Comment by ScopeyNZ** [comment]
Hmmmn interesting find. I'm not sure what kind of change this would be in regards to Semver... Someone could be relying on the cookie values.

**Comment by robbieaverill** [comment]
It might be more of a side effect of the public API than a part of it. I'm fine with this change being a patch

**Comment by robbieaverill** [comment]
Fixed in #264
